apps/bf_temp.c

```diff
@@ -0,0 +1,287 @@
+            || !BIO_meth_set_puts(methods_tempfile, tempfile_puts)
+            || !BIO_meth_set_gets(methods_tempfile, tempfile_gets)
+            || !BIO_meth_set_ctrl(methods_tempfile, tempfile_ctrl)
+            || !BIO_meth_set_create(methods_tempfile, tempfile_new)
+            || !BIO_meth_set_destroy(methods_tempfile, tempfile_free))
+            return NULL;
+    }
+    return methods_tempfile;
+}
+
+void apps_bf_tempfile_cleanup(void)
+{
+    BIO_meth_free(methods_tempfile);
+    methods_tempfile = NULL;
+}
+
+static int tempfile_new(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    temp = app_malloc(sizeof(*temp), "temp file");
+    temp->filename = NULL;
+    temp->tmp_filename = NULL;
+    BIO_set_data(b, temp);
+    BIO_set_init(b, 1);
+
+    return 1;
+}
+
+#ifdef _WIN32
+static int WIN32_rename(const char *from, const char *to);
+# define rename(from,to) WIN32_rename((from),(to))
+#endif
+
+static int tempfile_free(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    if (b == NULL)
+        return 0;
+
+    temp = BIO_get_data(b);
+    if (temp->filename != NULL && temp->tmp_filename != NULL) {
+#if !defined(__VMS)
+        rename(temp->tmp_filename, temp->filename);
+#endif
+    }
+    if (temp->filename)
+        OPENSSL_free(temp->filename);
+    if (temp->tmp_filename)
+        OPENSSL_free(temp->tmp_filename);
+    OPENSSL_free(temp);
+    BIO_set_data(b, NULL);
+    BIO_set_init(b, 0);
+
+    return 1;
+}
+
+static int tempfile_read(BIO *b, char *out, int outl)
+{
+    int ret = 0;
+    BIO *next = BIO_next(b);
+
+    if (out == NULL || outl == 0)
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    ret = BIO_read(next, out, outl);
+    return ret;
+}
+
+static int tempfile_write(BIO *b, const char *in, int inl)
+{
+    BIO *next = BIO_next(b);
+
+    if ((in == NULL) || (inl <= 0))
+        return 0;
+    if (next == NULL)
+        return 0;
+
+    return BIO_write(next, in, inl);
+}
+
+#define RND_SIZE 16
+
+static void make_temp_filename(const char *orig, char *filename)
+{
+    int i, size = RND_SIZE;
+    char *p;
+    char set[] = "0123456789abcdefghijklmnopqrstuvwxyz"
+                 "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    strcpy(filename, orig);
+    p = filename + strlen(orig);
+    *p++ = '.';
+
+    srand(time(NULL));
+    while (size--) {
+        i = rand() % (sizeof(set) - 1);
+        *p++ = set[i];
+    }
+
+    *p = '\0';
+
+    return;
+}
+
+static int istext(int format)
+{
+    return (format & B_FORMAT_TEXT) == B_FORMAT_TEXT;
+}
+
+static const char *modestr(char mode, int format)
+{
+    OPENSSL_assert(mode == 'a' || mode == 'r' || mode == 'w');
+
+    switch (mode) {
+    case 'a':
+        return istext(format) ? "a" : "ab";
+    case 'r':
+        return istext(format) ? "r" : "rb";
+    case 'w':
+        return istext(format) ? "w" : "wb";
+    }
+    /* The assert above should make sure we never reach this point */
+    return NULL;
+}
```

Comment:
Ugh, code copying :-/
Make them non-static in apps.c instead, declare them in apps.h, and use them instead of copying them

Reply:
Fixed